### ISSUE TITLE
HTTPResponse does not observe file-like object protocol if decode_content=True

### ISSUE THREAD

**Issue by hheimbuerger** [urllib3/urllib3#437]
My understanding was that urllib3's `HTTPResponse` is supposed to behave according to the file-like object protocol if `preload_content=False`. In particular, I would like to hand it over to another library that is able to read from a file-like object (e.g. lxml).

Unfortunately, the `HTTPReponse` does not appear to implement the protocol properly if `decode_content=True`. Test case:

```
import urllib3
print('urllib3 v{}'.format(urllib3.__version__))

http = urllib3.PoolManager()

response = http.request('GET', 'https://github.com',
                        headers={'Accept-Encoding': 'gzip'},
                        preload_content=False,
                        decode_content=False)
print('decode_content=False, read(1): expected 1 bytes, got {} bytes'.format(len(response.read(1))))
print('decode_content=False, read(500): expected 500 bytes, got {} bytes'.format(len(response.read(500))))

response = http.request('GET', 'https://github.com',
                        headers={'Accept-Encoding': 'gzip'},
                        preload_content=False,
                        decode_content=True)
print('decode_content=True, read(1): expected 1 bytes, got {} bytes'.format(len(response.read(1))))
print('decode_content=True, read(500): expected 500 bytes, got {} bytes'.format(len(response.read(500))))
```

Output:

```
urllib3 v1.9
decode_content=False, read(1): expected 1 bytes, got 1 bytes
decode_content=False, read(500): expected 500 bytes, got 500 bytes
decode_content=True, read(1): expected 1 bytes, got 0 bytes
decode_content=True, read(500): expected 500 bytes, got 1056 bytes
```

The problem with the first case is that I thought `.read()==''` means that the stream is consumed and/or has been closed. So a library might read 50 bytes, get an empty stream back and assume it has read the entire 'file', which wouldn't be the case with `HTTPResponse`.

Now you can argue that reading just one byte is unreasonable, but what's the minimal chunk size that still works? It appears to be 84 bytes for github.com, but it differs for other hosts.

The problem with the second case is that the file-object protocol defines the first parameter of `read()` as "[Read at most size bytes from the file](https://docs.python.org/2/library/stdtypes.html?highlight=read#file.read)". `HTTPResponse` returns more however. lxml for example seems to have internal buffer overflows if `read()` returns more bytes than it requested.

I guess at the very least, the documentation should mention that `HTTPResponse` is not adhering to the file-like object protocol.
##

**Comment by julian-r** [comment]
+1 to this discussion! 
Just to remark: It is really quite confusing that `read()` does not behave like a file read, which [says](https://docs.python.org/2/library/io.html#io.RawIOBase.read)

> Fewer than n bytes may be returned if the operating system call returns fewer than n bytes.

Should I add that to the documentation and do a merge request? (And I think it should implemented as a fix as well)

**Comment by Lukasa** [comment]
I'd honestly rather we fixed it than wrote a doc fix that we then have to back out once fixed. If anyone is planning to work on the fix then let's hold off. Otherwise, a doc change is acceptable. ;)

**Comment by julian-r** [comment]
In general the object returned by [`gzip.open`](https://docs.python.org/3/library/gzip.html) would support exactly the expected behavior.

The easiest way would be to wrap the body into a `gzip.open` in [response.py:138](https://github.com/shazow/urllib3/blob/master/urllib3/response.py#L138). But then the `decode_content` member won't work. Another idea I have is lazy wrapping `self._fp` depending on the `decode_content` member. Thoughts?

**Comment by Lukasa** [comment]
I think the better approach is always going to be to implement our own buffering, probably via wrapping `self._fp`. The key note is that `self._fp` isn't required to have anything but a `read` method.

**Comment by julian-r** [comment]
> The key note is that self._fp isn't required to have anything but a read method.

Which would be enough for `gzip.GzipFile`. I just tested that and it works fine for gzip streams. Not sure about deflate.

**Comment by Lukasa** [comment]
We would need it to be able to work for the three cases we currently tolerate in urllib3's code if we were going to use it. A quick glance at the API suggests it probably won't work in those cases. =(

**Comment by julian-r** [comment]
Yeah, with `length_remaining`,  and the `decode_content` handling with a non-file body a compatible solution seems to be quite complicated. What do you think about adding an property like `decoded_filestream` and a proper documentation into the existing `read ` function?

**Comment by Lukasa** [comment]
What do you envision those things looking like and doing?

**Comment by julian-r** [comment]
I mean there is no way to keep to implement transparent decompression while maintaining the compatibility of the `Response` object. Which behaves in combination with compression strange anyways because all the `amt` values are somehow "ignored" (also if we look at the `stream()` method).

So 2 options:

 1. break compatibility with `read`, `tell`, `stream`, `readinto` (which is broken anyway). I have no plan if this is going to make people happy or unhappy, or if anybody except a few will even notice.
 2. create a lazy property similar to the `stream()` method, returning a `io` compatible object, which wraps the `Request` object.

**Comment by Lukasa** [comment]
I don't entirely understand why you think there's no way to implement transparent decompression while maintaining compatibility. Can we not just internally buffer excess data?

**Comment by julian-r** [comment]
`tell()` would behave differently. Now it will return the actual bytes read from the underlying object. To behave in a `io` compatible manner, it would need to return the count of the data actually read.

**Comment by Lukasa** [comment]
I don't think that requirement exists. In particular, given that the response is *not* seekable, the value of `tell` cannot actually be fed into anything. Certainly the `io` module doesn't say that this is what `tell` has to return, and in plenty of cases it does not (e.g. for anything that implements `TextIOBase` `tell` is *explicitly* allowed to return an opaque value that needn't correspond to anything in the underlying bytestream).

**Comment by julian-r** [comment]
Ack. This will then change the following unittests:
 - `TestResponse.test_chunked_decoding_gzip`
 - `TestResponse.test_chunked_decoding_deflate2`
 - `TestResponse.test_chunked_decoding_deflate`.
 - `TestResponse.test_gzipped_streaming`
 - `TestResponse.test_deflate_streaming_tell_intermediate_point`
 - `TestResponse.test_deflate_streaming`
 - `TestResponse.test_deflate2_streaming`
Ack?


Just something I don't understand: What happens if the `transfer-encoding` is "chunked"? I think the read method then will return the whole chunked encoding *not* removing the "\r\n" and length headers, right? (I can't find a query to `self.chunked` in the read method)

**Comment by Lukasa** [comment]
@julian-r If `transfer-encoding` is chunked, the read method swaps over and calls `read_chunked` which does strip the chunked framing.

I'm happy to have those tests get changed. =)

**Comment by julian-r** [comment]
@Lukasa After staring a while at the code I can't find that reference to `read_chunked()` anyhow related to `read()`. I mean yes for `stream()`, but not for `read()`. Maybe I am blind, but could just give me a short hint?

**Comment by Lukasa** [comment]
Sorry, I wasn't clear enough: `httplib` has a `read_chunked` method that is called into automatically by `read` if needed.

**Comment by julian-r** [comment]
@Lukasa Ah I see. On py3 it is in `http.client.HTTPResponse`, There they implement also the readinto method quite clean for chunked messages. Now why do we have another `read_chunked` in the `urllib3.response.HTTPResponse` object as well? This way the response will able to handle chunked messages on a "not httplib `self._fp`" only if you call `stream` or `read_chunked` directly.

**Comment by Lukasa** [comment]
@julian-r We have a `read_chunked` in urllib3 to explicitly handle the desire to iterate over chunks with respect to their chunk boundaries.

**Comment by Lukasa** [comment]
Also, interestingly, to tolerate the use-case you just brought up (namely: non-httplib file objects).

**Comment by lbt** [comment]
Is this still active? It's certainly still seems to be a problem when trying to move to urllib3 and the "Backwards-compatible with http.client.HTTPResponse" (quote from the docs) can't be passed to libraries that took http.client.HTTPResponse because of this issue.

**Comment by sethmlarson** [comment]
@lbt see https://github.com/urllib3/urllib3/issues/2128
